### PR TITLE
BUG: allow for precision > 17 in longdouble repr test

### PR DESCRIPTION
**Pull request by matthew-brett**
Precision of repr print unconditionally set to 20, for longdoubles != double.  Allow for this
in repr round-trip test for longdouble.

Closes gh-7940.

### CODE REVIEW

File: numpy/core/tests/test_longdouble.py [hunk 1]
```diff
@@ -30,8 +30,11 @@ def test_scalar_extraction():
...
 
 def test_repr_roundtrip():
-    o = 1 + np.finfo(np.longdouble).eps
-    assert_equal(np.longdouble(repr(o)), o,
+    o = 1 + LD_INFO.eps
+    # We will only see eps in repr if within printing precision.
+    # Printing precision defined as 17 in scalartypes.c.src
+    expected = 1.0 if LD_INFO.precision > 17 else o
+    assert_equal(np.longdouble(repr(o)), expected,
                  "repr was %s" % repr(o))
```
Comment (eric-wieser) [inline numpy/core/tests/test_longdouble.py]: Wait, isn't the bug here that the repr is capped at a precision of `17` when it shouldn't be?

Comment (eric-wieser) [inline numpy/core/tests/test_longdouble.py]: Also, `scalartypes.c.src` seems to cap it at `20`, not 17, unless `longdouble` and `double` are the same

Comment (matthew-brett) [inline numpy/core/tests/test_longdouble.py]: Yes, you're right, I was just looking at that code.  Can correct the comment and constant, but the question is, do we really want - for example - 31 digits of precision for np.longdouble (that's what double double gives, on PPC64).  If so, we'd have to adapt the scalartypes.c.src code to detect the different types of longdouble.

Comment (eric-wieser) [inline numpy/core/tests/test_longdouble.py]: I think it's desirable that `repr` roundtrips properly, and I don't think that changing the test to allow the failure here is the right solution. So yes, on PPC64 I think that it should print 31 digits for repr - there's always `str` for the sane-length output.

Obviously it's not desirable use 31 digits on platforms that don't support it, but as a quick fix, I think that 20 should be become 33 (precision +2 used elsewhere).

Adapting `scalartypes.c.src` to choose more intelligently is an obvious next step.

**Comment by matthew-brett** [comment]
Pointing to now-hidden discussion : https://github.com/numpy/numpy/pull/8656#discussion_r102326216

The question is - how much do we care about the round-trippiness of repr for longdouble?   If we do care about that then:

a) we'd get 33 digits of repr precision on PPC64, and 36 on SPARC (real float128) and
b) we'd have to re-implement the longdouble type detection in scalartype.c.src - which might be easy, but I don't know how to do it.

**Comment by eric-wieser** [comment]
Side-note: make a PR with just the `LD_INFO` changes and I'll merge it

**Comment by matthew-brett** [comment]
My vote would be to give up on struct repr roundtrip correctness, at least for now, because:

* we haven't previously had this for np.longdouble not in (float64, float80) and;
* 33 digits will be a lot of text for values in an array and;
* changing the repr for longdouble might disturb some doctests and;
* it appears from the top of `numpy/core/tests/test_longdouble.py` that some platforms lack (an accurate?) `strtold_l` and therefore repr roundtrip cannot work for these, and;
* this workaround already seems better than the failing test.

**Comment by eric-wieser** [comment]
> 33 digits will be a lot of text for values in an array 

As far as I remember, `ndarray.__repr__` does not directly invoke `self.dtype.type.__repr__`, but instead uses the (independent) print settings. So I don't think this argument holds

**Comment by matthew-brett** [comment]
OK - happy to withdraw that part of the argument.

**Comment by eric-wieser** [comment]
I'm not convinced the workaround is correct though. I think that if the precision is too high, the test should be skipped altogether, rather than verifying that it does _not_ roundtrip

**Comment by matthew-brett** [comment]
How about:

```
def test_repr_roundtrip():
    o = 1 + LD_INFO.eps
    # We will only see eps in repr if within printing precision.
    repr_precision = len(repr(np.longdouble(0.1))) - 2
    if LD_INFO.precision >= repr_precision:
        assert_equal(np.longdouble(repr(o)), o, "repr was %s" % repr(o))
```

Then, if we want to increase the repr precision later, the test will be triggered automatically.

**Comment by eric-wieser** [comment]
That sounds pretty good to me, although I might still be tempted to mark it as a skipped test somehow rather than it no-opping

**Comment by matthew-brett** [comment]
OK - I pushed up a version where we skip the test when there isn't enough repr precision.

File: numpy/core/tests/test_longdouble.py [hunk 2]
```diff
@@ -10,29 +10,31 @@
...
-    o = 1 + np.finfo(np.longdouble).eps
+    o = 1 + LD_INFO.eps
     a = np.array([o, o, o])
     assert_equal(a[1], o)
 
 
 # Conversions string -> long double
 
 
+@dec.skipif(LD_INFO.precision >= len(repr(np.longdouble(0.1))) - 2,
```
Comment (eric-wieser) [inline numpy/core/tests/test_longdouble.py]: Probably better computing `len(repr(np.longdouble(0.1)))` on another line, doing `LD_INFO.precision + 2 >= repr_precision`, and adding a comment linking that `+2` to `scalartypes`

**Comment by matthew-brett** [comment]
Made that change.  Also (sorry) - I re-enabled an unrelated test that was being overwritten.

**Comment by matthew-brett** [comment]
`test_longdouble` tests pass from this branch on SPARC and PPC64el.

**Comment by eric-wieser** [comment]
Looks good to me. I doubt anyone is actually going to be able to confirm those test results on SPARC or PPC64el, so I'll take your word for it

**Pull request merged**